{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install pyBigWig "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install pybbi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### [BigWig file format](https://genome.ucsc.edu/goldenPath/help/bigWig.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyBigWig as bw\n",
    "import bbi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapp_path = \"/mnt/beegfs/userdata/j_wang/Algorithm_Resources/dig/Mappability/wgEncodeCrgMapabilityAlign36mer.bigWig\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapp_fd = bbi.open(mapp_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on BBIFile object:\n",
      "\n",
      "class BBIFile(builtins.object)\n",
      " |  Interface to a UCSC Big Binary Indexed (BBI) file.\n",
      " |  \n",
      " |  The resource may be a bigWig or a bigBed file.\n",
      " |  BigBed AutoSql schemas are supported.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __enter__(self)\n",
      " |      BBIFile.__enter__(self)\n",
      " |  \n",
      " |  __exit__(self, typ, val, tb)\n",
      " |      BBIFile.__exit__(self, typ, val, tb)\n",
      " |  \n",
      " |  __reduce__ = __reduce_cython__(...)\n",
      " |      BBIFile.__reduce_cython__(self)\n",
      " |  \n",
      " |  __setstate__ = __setstate_cython__(...)\n",
      " |      BBIFile.__setstate_cython__(self, __pyx_state)\n",
      " |  \n",
      " |  best_zoom(self, chrom, start, end, bins)\n",
      " |      BBIFile.best_zoom(self, str chrom, int start, int end, int bins)\n",
      " |  \n",
      " |  close(self)\n",
      " |      BBIFile.close(self)\n",
      " |  \n",
      " |  fetch(self, chrom, start, end, bins=-1, missing=0.0, oob=nan, summary='mean', exact=False)\n",
      " |      BBIFile.fetch(self, str chrom, int start, int end, int bins=-1, double missing=0.0, double oob=np.nan, str summary='mean', bool exact=False)\n",
      " |      \n",
      " |      Read the signal data in a BBI file overlapping a genomic query interval\n",
      " |      into a numpy array.\n",
      " |      \n",
      " |      If a number of bins is requested, this will return a summary statistic\n",
      " |      for each bin. Otherwise, the data is returne unsummarized at base pair\n",
      " |      resolution (default).\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      chrom : str\n",
      " |          Chromosome name.\n",
      " |      start : int\n",
      " |          Start coordinate. If start is less than zero, the beginning of the\n",
      " |          track is not truncated but treated as out of bounds.\n",
      " |      end : int\n",
      " |          End coordinate. If end is less than zero, the end is set to the\n",
      " |          chromosome size. If end is greater than the chromosome size, the\n",
      " |          end of the track is not truncated but treated as out of bounds.\n",
      " |      bins : int, optional\n",
      " |          Number of bins to divide the query interval into for coarsegraining.\n",
      " |          Default (-1) means no summarization (i.e., 1 bp bins).\n",
      " |      missing : float, optional [default: 0.0]\n",
      " |          Fill-in value for unreported data in valid regions. Default is 0.\n",
      " |      oob : float, optional [default: NaN]\n",
      " |          Fill-in value for out-of-bounds regions. Default is NaN.\n",
      " |      summary : str, optional [default: 'mean']\n",
      " |          Summary statistic to use if ``bins`` is specified. Options are\n",
      " |          'mean', 'sum', 'min', 'max', 'cov' (coverage), and 'std'\n",
      " |          (standard deviation). Default is 'mean'.\n",
      " |      exact : bool, optional [default: False]\n",
      " |          If True and ``bins`` is specified, return exact summary statistic\n",
      " |          values instead of interpolating from the optimal zoom level.\n",
      " |          Default is False.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      1D ndarray\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Values returned for a BigWig file are derived from its quantitative\n",
      " |      \"value\" field. A BigWig file encodes a step function, and the value at\n",
      " |      a base is given by the signal value of the unique interval that\n",
      " |      contains that base.\n",
      " |      \n",
      " |      Values returned for a BigBed file are derived from the coverage\n",
      " |      (i.e., pileup) of the intervals. A BigBed file encodes a collection\n",
      " |      of (possibly overlapping) intervals which may or may not be associated\n",
      " |      with quantitative scores. The \"value\" at given base used here\n",
      " |      summarizes the number of intervals overlapping that base, not a score.\n",
      " |      \n",
      " |      If a number of bins is requested and ``exact`` is False, the summary\n",
      " |      data is interpolated from the closest available zoom level. If you\n",
      " |      need accurate summary data and are okay with small trade-off in speed,\n",
      " |      set ``exact`` to True.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      stackup : Stack signal tracks from many query intervals into a matrix\n",
      " |  \n",
      " |  fetch_intervals(self, chrom, start, end, iterator=False)\n",
      " |      BBIFile.fetch_intervals(self, str chrom, int start, int end, bool iterator=False)\n",
      " |      \n",
      " |      Return an iterator or data frame of feature intervals overlapping a\n",
      " |      specified genomic query interval.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      chrom : str\n",
      " |          Chromosome name.\n",
      " |      start : int\n",
      " |          Start coordinate.\n",
      " |      end : int\n",
      " |          End coordinate. If end is less than zero, the end is set to the\n",
      " |          chromosome size.\n",
      " |      iterator : bool, optional\n",
      " |          If True, return an iterator that provides records as tuples. For\n",
      " |          bigBeds, extra BED fields will be returned as unparsed strings. If\n",
      " |          False, return a DataFrame with all columns parsed according to the\n",
      " |          file's autosql schema.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      pandas.DataFrame or BigWigIntervalIterator or BigBedIntervalIterator\n",
      " |  \n",
      " |  fetch_summaries(self, chrom, start, end, zoom=0)\n",
      " |      BBIFile.fetch_summaries(self, str chrom, int start, int end, int zoom=0)\n",
      " |      \n",
      " |      Return a DataFrame of complete summary statistic bins overlapping a\n",
      " |      genomic query interval at a specific zoom level.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      chrom : str\n",
      " |          Chromosome name.\n",
      " |      start : int\n",
      " |          Start coordinate.\n",
      " |      end : int\n",
      " |          End coordinate. If end is less than zero, the end is set to the\n",
      " |          chromosome size.\n",
      " |      zoom : int, optional\n",
      " |          Zoom level. Default is 0.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      pandas.DataFrame\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      The summary fields are 'validCount', 'min', 'max', 'sum', 'sumSquares'.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      fetch : Fetch the signal track of a single interval\n",
      " |      fetch_intervals : Fetch feature intervals overlapping a genomic interval\n",
      " |  \n",
      " |  read_autosql(self)\n",
      " |      BBIFile.read_autosql(self)\n",
      " |  \n",
      " |  stackup(self, chroms, starts, ends, bins=-1, missing=0.0, oob=nan, summary='mean', exact=False)\n",
      " |      BBIFile.stackup(self, chroms, starts, ends, int bins=-1, double missing=0.0, double oob=np.nan, str summary='mean', bool exact=False)\n",
      " |      \n",
      " |      Vertically stack signal tracks from many query intervals into a matrix.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      chroms : array-like of str\n",
      " |          Chromosome names of the query intervals.\n",
      " |      starts : array-like of int\n",
      " |          Start coordinates. If start is less than zero, the beginning of the\n",
      " |          track is not truncated but treated as out of bounds.\n",
      " |      ends : array-like of int\n",
      " |          End coordinates. If end is less than zero, the end is set to the\n",
      " |          chromosome size. If end is greater than the chromosome size, the\n",
      " |          end of the track is not truncated but treated as out of bounds.\n",
      " |      bins : int\n",
      " |          Number of bins to summarize the data. Default (-1) means no\n",
      " |          aggregation: each bin is 1 base pair and the query intervals must\n",
      " |          have the same length. If bins > 0, the input query intervals can\n",
      " |          have different lengths, and the data will be rescaled to the same\n",
      " |          number of bins.\n",
      " |      missing : float\n",
      " |          Fill-in value for unreported data in valid regions. Default is 0.\n",
      " |      oob : float\n",
      " |          Fill-in value for out-of-bounds regions. Default is NaN.\n",
      " |      summary : str, optional\n",
      " |          Summary statistic to use if summarizing. Options are 'mean', 'min',\n",
      " |          'max', 'cov' (coverage), and 'std' (standard deviation). Default is\n",
      " |          'mean'.\n",
      " |      exact : bool, optional [default: False]\n",
      " |          If True and ``bins`` is specified, return exact summary statistic\n",
      " |          values instead of interpolating from the optimal zoom level.\n",
      " |          Default is False.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      out : 2D ndarray of float64 (n_intervals, n_bins)\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      If not summarizing, the input intervals must have the same length.\n",
      " |      \n",
      " |      If a number of bins is requested, intervals may have variable lengths\n",
      " |      and the data from each interval will be rescaled to the same number\n",
      " |      of bins (e.g., aligning gene bodies).\n",
      " |      \n",
      " |      If a number of bins is requested and ``exact`` is False, the summary\n",
      " |      data is interpolated from the closest available zoom level. If you\n",
      " |      need accurate summary data and are okay with small trade-off in\n",
      " |      speed, set ``exact`` to True.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      fetch : Fetch the signal track of a single interval\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  chromsizes\n",
      " |      An ordered dictionary of chromosome names mapped to their sizes in bp.\n",
      " |  \n",
      " |  closed\n",
      " |  \n",
      " |  info\n",
      " |      A dict of information about the bbi file.\n",
      " |  \n",
      " |  is_bigbed\n",
      " |  \n",
      " |  is_bigwig\n",
      " |  \n",
      " |  is_remote\n",
      " |  \n",
      " |  path\n",
      " |  \n",
      " |  schema\n",
      " |  \n",
      " |  zooms\n",
      " |      A list of \"reduction levels\" (bin sizes), i.e. the number of bases per\n",
      " |      summary item.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mapp_fd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'bedGraph',\n",
       " 'comment': '',\n",
       " 'columns': ['chrom', 'start', 'end', 'value'],\n",
       " 'dtypes': {'chrom': 'object',\n",
       "  'start': 'uint32',\n",
       "  'end': 'uint32',\n",
       "  'value': 'float32'},\n",
       " 'description': {'chrom': 'Reference sequence chromosome or scaffold',\n",
       "  'start': 'Start position in chromosome',\n",
       "  'end': 'End position in chromosome',\n",
       "  'value': 'Data value'}}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapp_fd.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': 249250621,\n",
       " 'chr10': 135534747,\n",
       " 'chr11': 135006516,\n",
       " 'chr12': 133851895,\n",
       " 'chr13': 115169878,\n",
       " 'chr14': 107349540,\n",
       " 'chr15': 102531392,\n",
       " 'chr16': 90354753,\n",
       " 'chr17': 81195210,\n",
       " 'chr18': 78077248,\n",
       " 'chr19': 59128983,\n",
       " 'chr2': 243199373,\n",
       " 'chr20': 63025520,\n",
       " 'chr21': 48129895,\n",
       " 'chr22': 51304566,\n",
       " 'chr3': 198022430,\n",
       " 'chr4': 191154276,\n",
       " 'chr5': 180915260,\n",
       " 'chr6': 171115067,\n",
       " 'chr7': 159138663,\n",
       " 'chr8': 146364022,\n",
       " 'chr9': 141213431,\n",
       " 'chrM': 16571,\n",
       " 'chrX': 155270560,\n",
       " 'chrY': 59373566}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapp_fd.chromsizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'version': 4,\n",
       " 'isCompressed': True,\n",
       " 'isSwapped': 0,\n",
       " 'primaryDataSize': 1174714010,\n",
       " 'zoomLevels': 10,\n",
       " 'chromCount': 25,\n",
       " 'summary': {'basesCovered': 2861332409,\n",
       "  'sum': 2258139576.414049,\n",
       "  'mean': 0.78919162601008,\n",
       "  'min': 2.265667717438191e-06,\n",
       "  'max': 1.0,\n",
       "  'std': 0.38573641381908746}}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapp_fd.info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method best_zoom in module bbi.cbbi:\n",
      "\n",
      "best_zoom(chrom, start, end, bins) method of bbi.cbbi.BBIFile instance\n",
      "    BBIFile.best_zoom(self, str chrom, int start, int end, int bins)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mapp_fd.best_zoom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method stackup in module bbi.cbbi:\n",
      "\n",
      "stackup(chroms, starts, ends, bins=-1, missing=0.0, oob=nan, summary='mean', exact=False) method of bbi.cbbi.BBIFile instance\n",
      "    BBIFile.stackup(self, chroms, starts, ends, int bins=-1, double missing=0.0, double oob=np.nan, str summary='mean', bool exact=False)\n",
      "    \n",
      "    Vertically stack signal tracks from many query intervals into a matrix.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    chroms : array-like of str\n",
      "        Chromosome names of the query intervals.\n",
      "    starts : array-like of int\n",
      "        Start coordinates. If start is less than zero, the beginning of the\n",
      "        track is not truncated but treated as out of bounds.\n",
      "    ends : array-like of int\n",
      "        End coordinates. If end is less than zero, the end is set to the\n",
      "        chromosome size. If end is greater than the chromosome size, the\n",
      "        end of the track is not truncated but treated as out of bounds.\n",
      "    bins : int\n",
      "        Number of bins to summarize the data. Default (-1) means no\n",
      "        aggregation: each bin is 1 base pair and the query intervals must\n",
      "        have the same length. If bins > 0, the input query intervals can\n",
      "        have different lengths, and the data will be rescaled to the same\n",
      "        number of bins.\n",
      "    missing : float\n",
      "        Fill-in value for unreported data in valid regions. Default is 0.\n",
      "    oob : float\n",
      "        Fill-in value for out-of-bounds regions. Default is NaN.\n",
      "    summary : str, optional\n",
      "        Summary statistic to use if summarizing. Options are 'mean', 'min',\n",
      "        'max', 'cov' (coverage), and 'std' (standard deviation). Default is\n",
      "        'mean'.\n",
      "    exact : bool, optional [default: False]\n",
      "        If True and ``bins`` is specified, return exact summary statistic\n",
      "        values instead of interpolating from the optimal zoom level.\n",
      "        Default is False.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : 2D ndarray of float64 (n_intervals, n_bins)\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    If not summarizing, the input intervals must have the same length.\n",
      "    \n",
      "    If a number of bins is requested, intervals may have variable lengths\n",
      "    and the data from each interval will be rescaled to the same number\n",
      "    of bins (e.g., aligning gene bodies).\n",
      "    \n",
      "    If a number of bins is requested and ``exact`` is False, the summary\n",
      "    data is interpolated from the closest available zoom level. If you\n",
      "    need accurate summary data and are okay with small trade-off in\n",
      "    speed, set ``exact`` to True.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    fetch : Fetch the signal track of a single interval\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mapp_fd.stackup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method fetch in module bbi.cbbi:\n",
      "\n",
      "fetch(chrom, start, end, bins=-1, missing=0.0, oob=nan, summary='mean', exact=False) method of bbi.cbbi.BBIFile instance\n",
      "    BBIFile.fetch(self, str chrom, int start, int end, int bins=-1, double missing=0.0, double oob=np.nan, str summary='mean', bool exact=False)\n",
      "    \n",
      "    Read the signal data in a BBI file overlapping a genomic query interval\n",
      "    into a numpy array.\n",
      "    \n",
      "    If a number of bins is requested, this will return a summary statistic\n",
      "    for each bin. Otherwise, the data is returne unsummarized at base pair\n",
      "    resolution (default).\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    chrom : str\n",
      "        Chromosome name.\n",
      "    start : int\n",
      "        Start coordinate. If start is less than zero, the beginning of the\n",
      "        track is not truncated but treated as out of bounds.\n",
      "    end : int\n",
      "        End coordinate. If end is less than zero, the end is set to the\n",
      "        chromosome size. If end is greater than the chromosome size, the\n",
      "        end of the track is not truncated but treated as out of bounds.\n",
      "    bins : int, optional\n",
      "        Number of bins to divide the query interval into for coarsegraining.\n",
      "        Default (-1) means no summarization (i.e., 1 bp bins).\n",
      "    missing : float, optional [default: 0.0]\n",
      "        Fill-in value for unreported data in valid regions. Default is 0.\n",
      "    oob : float, optional [default: NaN]\n",
      "        Fill-in value for out-of-bounds regions. Default is NaN.\n",
      "    summary : str, optional [default: 'mean']\n",
      "        Summary statistic to use if ``bins`` is specified. Options are\n",
      "        'mean', 'sum', 'min', 'max', 'cov' (coverage), and 'std'\n",
      "        (standard deviation). Default is 'mean'.\n",
      "    exact : bool, optional [default: False]\n",
      "        If True and ``bins`` is specified, return exact summary statistic\n",
      "        values instead of interpolating from the optimal zoom level.\n",
      "        Default is False.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    1D ndarray\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    Values returned for a BigWig file are derived from its quantitative\n",
      "    \"value\" field. A BigWig file encodes a step function, and the value at\n",
      "    a base is given by the signal value of the unique interval that\n",
      "    contains that base.\n",
      "    \n",
      "    Values returned for a BigBed file are derived from the coverage\n",
      "    (i.e., pileup) of the intervals. A BigBed file encodes a collection\n",
      "    of (possibly overlapping) intervals which may or may not be associated\n",
      "    with quantitative scores. The \"value\" at given base used here\n",
      "    summarizes the number of intervals overlapping that base, not a score.\n",
      "    \n",
      "    If a number of bins is requested and ``exact`` is False, the summary\n",
      "    data is interpolated from the closest available zoom level. If you\n",
      "    need accurate summary data and are okay with small trade-off in speed,\n",
      "    set ``exact`` to True.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    stackup : Stack signal tracks from many query intervals into a matrix\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mapp_fd.fetch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method fetch_intervals in module bbi.cbbi:\n",
      "\n",
      "fetch_intervals(chrom, start, end, iterator=False) method of bbi.cbbi.BBIFile instance\n",
      "    BBIFile.fetch_intervals(self, str chrom, int start, int end, bool iterator=False)\n",
      "    \n",
      "    Return an iterator or data frame of feature intervals overlapping a\n",
      "    specified genomic query interval.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    chrom : str\n",
      "        Chromosome name.\n",
      "    start : int\n",
      "        Start coordinate.\n",
      "    end : int\n",
      "        End coordinate. If end is less than zero, the end is set to the\n",
      "        chromosome size.\n",
      "    iterator : bool, optional\n",
      "        If True, return an iterator that provides records as tuples. For\n",
      "        bigBeds, extra BED fields will be returned as unparsed strings. If\n",
      "        False, return a DataFrame with all columns parsed according to the\n",
      "        file's autosql schema.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    pandas.DataFrame or BigWigIntervalIterator or BigBedIntervalIterator\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mapp_fd.fetch_intervals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10078</td>\n",
       "      <td>0.001362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10078</td>\n",
       "      <td>10081</td>\n",
       "      <td>0.023810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10081</td>\n",
       "      <td>10088</td>\n",
       "      <td>0.018519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10088</td>\n",
       "      <td>10089</td>\n",
       "      <td>0.014706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10089</td>\n",
       "      <td>10096</td>\n",
       "      <td>0.018519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5707</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99368</td>\n",
       "      <td>99369</td>\n",
       "      <td>0.018519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5708</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99369</td>\n",
       "      <td>99370</td>\n",
       "      <td>0.038462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5709</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99370</td>\n",
       "      <td>99371</td>\n",
       "      <td>0.047619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5710</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99371</td>\n",
       "      <td>99375</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5711</th>\n",
       "      <td>chr1</td>\n",
       "      <td>99375</td>\n",
       "      <td>100000</td>\n",
       "      <td>0.076923</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5712 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  start     end     value\n",
       "0     chr1  10000   10078  0.001362\n",
       "1     chr1  10078   10081  0.023810\n",
       "2     chr1  10081   10088  0.018519\n",
       "3     chr1  10088   10089  0.014706\n",
       "4     chr1  10089   10096  0.018519\n",
       "...    ...    ...     ...       ...\n",
       "5707  chr1  99368   99369  0.018519\n",
       "5708  chr1  99369   99370  0.038462\n",
       "5709  chr1  99370   99371  0.047619\n",
       "5710  chr1  99371   99375  0.062500\n",
       "5711  chr1  99375  100000  0.076923\n",
       "\n",
       "[5712 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapp_fd.fetch_intervals('chr1', 1, 100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RepDigDrive",
   "language": "python",
   "name": "repdigdrive"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
